bitcoin/src/address.rs

```diff
@@ -727,34 +865,34 @@ impl Address {
         }
     }
 
-    /// Returns true if the given pubkey is directly related to the address payload.
+    /// Checks whether network of this address is as required.
     ///
-    /// This is determined by directly comparing the address payload with either the
-    /// hash of the given public key or the segwit redeem hash generated from the
-    /// given key. For taproot addresses, the supplied key is assumed to be tweaked
-    pub fn is_related_to_pubkey(&self, pubkey: &PublicKey) -> bool {
-        let pubkey_hash = pubkey.pubkey_hash();
-        let payload = self.payload.as_bytes();
-        let xonly_pubkey = XOnlyPublicKey::from(pubkey.inner);
-
-        (*pubkey_hash.as_ref() == *payload)
-            || (xonly_pubkey.serialize() == *payload)
-            || (*segwit_redeem_hash(&pubkey_hash).as_ref() == *payload)
+    /// For details about this mechanism, see section [*Parsing addresses*](Address#parsing-addresses)
+    /// on [`Address`].
+    #[inline]
+    pub fn require_network(self, required: Network) -> Result<Address, Error> {
+        if self.is_valid_for_network(required) {
+            Ok(self.assume_checked())
+        } else {
+            Err(Error::NetworkValidation { found: self.network, required })
+        }
     }
 
-    /// Returns true if the supplied xonly public key can be used to derive the address.
+    /// Marks, without any additional checks, network of this address as checked.
     ///
-    /// This will only work for Taproot addresses. The Public Key is
-    /// assumed to have already been tweaked.
-    pub fn is_related_to_xonly_pubkey(&self, xonly_pubkey: &XOnlyPublicKey) -> bool {
-        let payload = self.payload.as_bytes();
-        payload == xonly_pubkey.serialize()
+    /// Improper use of this method may lead to loss of funds. Reader will most likely prefer
+    /// [`require_network`](Address<NetworkUnchecked>::require_network) as a safe variant.
+    /// For details about this mechanism, see section [*Parsing addresses*](Address#parsing-addresses)
+    /// on [`Address`].
+    #[inline]
+    pub fn assume_checked(self) -> Address {
+        Address::new(self.network, self.payload)
     }
 }
 
 // Alternate formatting `{:#}` is used to return uppercase version of bech32 addresses which should
 // be used in QR codes, see [`Address::to_qr_uri`].
-impl fmt::Display for Address {
+impl<V: NetworkValidation> fmt::Display for Address<V> {
```

Comment:
Is this here only because calling `Debug` is wrong and we want to avoid duplicating the code since debug is (and should be) implemented for both? Note that @apoelstra didn't address my argument that people who want to just validate can call `assume_valid`. Or display the original string which will be faster anyway.

I suggest to have `fmt_unchecked` private method and call it from both impls, `Display` being only available for checked.

Oh and maybe we should mess up debug formatting a bit to discourage invalid usage? (put `Address()` around it or something).

Reply:
Reasonable. I will address this in my next commit.

Reply:
This brings an issue with serde and I am not sure what would be the best way around it.

Currently, we have `Serialize` and `Deserialize` for `Address<NetworkUnchecked>` (via macro) and `Serialize` for `Address<NetworkChecked>` (manual impl).

`Serialize` for `Address<NetworkUnchecked>` uses `Display` (in the macro). After removing `Display` from `Address<NetworkUnchecked>`, it can't be used anymore.

I guess most correct way would be to have only `Deserialize` for `Address<NetworkUnchecked>` and `Serialize` for `Address<NetworkChecked>` but it seems to me it would hurt ergonomics of the usage.

Alternative is to keep `Serialize` for `Address<NetworkUnchecked>` and make it work (via private fmt).

What do people think would be better approach?

Reply:
I think the best approach is to use the private fmt method.

Reply:
> but it seems to me it would hurt ergonomics of the usage.

This might be true. If people have it inside a larger struct then it can get annoying since they have to move it out and in. Ideally they would have the struct tagged as well but there are n Rust-native tools to compose this. I will give it more thought.